### ISSUE TITLE
Locality based routing not working after Istio 1.19 upgrade

### ISSUE THREAD

**Issue by saansh45** [istio/istio#48302]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

Recently we upgraded Istio control plane from `1.18.5` to `1.19.3`. After the upgrade we noticed that locality based routing is not working. After much debugging, we found that our client was missing service object. Once we create client service, locality based routing started working. 

In `1.18.5`, we don't need to have client service dependency. Locality based routing is working without client service object in `1.18.5`. 

### Version

```Text
$ istioctl version
istiod version: 1.19.3
data plane version: 1.19.3 (24 proxies)

$ kubectl version
Client Version: v1.28.1
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.27.7-eks-4f4795d
```

### Additional Information

*Destination Rule*

```
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: helloworld-1-19
  namespace: istio-1-19-test
spec:
  host: helloworld-1-19
  trafficPolicy:
    loadBalancer:
      localityLbSetting:
        failoverPriority:
        - topology.istio.io/network
        - topology.kubernetes.io/region
        - topology.kubernetes.io/zone
        - topology.istio.io/subzone
    outlierDetection:
      baseEjectionTime: 3s
      consecutiveGatewayErrors: 10
      interval: 30s
      maxEjectionPercent: 30
      minHealthPercent: 50
```

*Pod details*

curlpods are clients and helloworld pods are our application services. 

```
$ k get pods -owide
NAME                               READY   STATUS    RESTARTS   AGE    IP               NODE                                           NOMINATED NODE   READINESS GATES
curlpod-1-19-74b9477f4b-7w654      2/2     Running   0          115s   100.64.58.232    ip-100-64-60-225.us-west-2.compute.internal    <none>           <none>
curlpod-1-19-74b9477f4b-brttx      2/2     Running   0          115s   100.64.76.12     ip-100-64-127-126.us-west-2.compute.internal   <none>           <none>
curlpod-1-19-74b9477f4b-fn85x      2/2     Running   0          115s   100.64.187.178   ip-100-64-190-43.us-west-2.compute.internal    <none>           <none>
helloworld-1-19-567f7f58cb-j6z8x   2/2     Running   0          115s   100.64.58.233    ip-100-64-60-225.us-west-2.compute.internal    <none>           <none>
helloworld-1-19-567f7f58cb-whk75   2/2     Running   0          115s   100.64.185.192   ip-100-64-190-43.us-west-2.compute.internal    <none>           <none>
helloworld-1-19-567f7f58cb-zwbcb   2/2     Running   0          115s   100.64.74.96     ip-100-64-127-126.us-west-2.compute.internal   <none>           <none>
```

*Service*

```
kubectl get svc -owide -n istio-1-19-test

NAME              TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)   AGE    SELECTOR
helloworld-1-19   ClusterIP   172.20.177.113   <none>        80/TCP    113m   app=helloworld-1-19
```

*Test with first client curlpod*

With locality based routing we were expecting the traffic from `curlpod-1-19-74b9477f4b-7w654` (running on `ip-100-64-60-225.us-west-2.compute.internal`) should go to the `helloworld-1-19-567f7f58cb-j6z8x` pod which is also running on the same node. But it was sending traffic to another pod. 

```
$ kubectl exec curlpod-1-19-74b9477f4b-7w654 -- /bin/sh -c "while true; do curl -s helloworld-1-19/hello; done"
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
```

*Test with second client curlpod*

`curlpod-1-19-74b9477f4b-fn85x` running on `ip-100-64-190-43.us-west-2.compute.internal` should send traffic to `helloworld-1-19-567f7f58cb-whk75` pod  which is running on same node. But it wasn't the case. 

```
$ kubectl exec curlpod-1-19-74b9477f4b-fn85x -- /bin/sh -c "while true; do curl -s helloworld-1-19/hello; done"
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
```

*Test with third client curlpod*

`curlpod-1-19-74b9477f4b-brttx`  running on `ip-100-64-127-126.us-west-2.compute.internal` was sending traffic `helloworld-1-19-567f7f58cb-zwbcb` pod running on the same node. 

```
$ kubectl exec curlpod-1-19-74b9477f4b-brttx -- /bin/sh -c "while true; do curl -s helloworld-1-19/hello; done"
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
Hello version: v2, instance: helloworld-1-19-567f7f58cb-zwbcb
```

If you notice here, istio-proxy is always sending traffic to the same pod from all proxies. Interesting thing here is it's the first entry from cluster endpoints. And Istio-proxy always sends traffic to the same proxy.  

As soon as we add client service and rollout the pods, locality based routing starts working in 1.19.3. This doesn't happen in Istio `1.18.5`. Currently we run several workloads and these clients don't have service object and locality based routing works fine. 

This looks like a bug in Istio 1.19 version.

**Comment by howardjohn** [comment]
This was only intentionally (afaik) fixed in 1.20 by https://github.com/istio/istio/pull/46295. Not sure how it accidentally got fixed in 1.18, then re-broken in 1.19 (it has been "broken" for ~ever). 

Note first line of https://istio.io/v1.19/docs/ops/deployment/requirements/#pod-requirements vs https://istio.io/latest/docs/ops/deployment/requirements/.

So overall, I wouldn't consider "locality LB requires client service in 1.19" to be a bug really. Given the behavior changes from 1.18 it may be a sign there is something else going on unexpected, though.

In any case, i would recommend upgrading to 1.20 personally

**Comment by hzxuzhonghu** [comment]
I doubt there is some regression on proxy labels, can you check if this is caused by missing locality labels in proxcy?

curl istiod /debug/adsz, it will list the labels
